### PR TITLE
Multiple memoryserver implementation phase 1

### PR DESCRIPTION
**Pull request by gautham-bhat-k**
1] Adding new Allocator and CIS classes

### CODE REVIEW

File: src/allocator/fam_allocator.h [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * fam_allocator.h
+ * fam_allocator_grpc.h
```
[the remark this answers is not in the document]
Comment (mashoodk) [reply src/allocator/fam_allocator.h:2]: Shouldnt we remove fam_allocator_nvmm and fam_allocator_grpc. Make sure nvmm specific things are covered. eg: fam_map/ cas128 etc.

File: src/allocator/fam_allocator_client.cpp [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    uint64_t regionId = globalDescriptor.regionId;
+    uint64_t offset = globalDescriptor.offset;
+    uint64_t memoryServerId = descriptor->get_memserver_id();
+    try {
+        return famCIS->fam_map(regionId, offset, memoryServerId, uid, gid);
+    } catch (Memserver_Exception &e) {
+        throw Fam_Allocator_Exception((enum Fam_Error)e.fam_error(),
+                                      e.fam_error_msg());
+    }
+} // namespace openfam
```
Comment (mashoodk) [inline src/allocator/fam_allocator_client.cpp:334]: remove "namespace openfam" comment

File: src/allocator/fam_allocator_nvmm.cpp [hunk 3]
```diff
@@ -56,8 +56,7 @@ Fam_Region_Descriptor *Fam_Allocator_NVMM::create_region(
     try {
         allocator->create_region(name, regionId, (size_t)nbytes, permissions,
                                  uid, gid);
-    }
```
Comment (mashoodk) [inline src/allocator/fam_allocator_nvmm.cpp:59]: Shouldnt we be deleting this file ?

File: src/cis/fam_cis.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
+/*
+ * fam_cis.h
+ * Copyright (c) 2019 Hewlett Packard Enterprise Development, LP. All rights
```
Comment (mashoodk) [inline src/cis/fam_cis.h:3]: Copyright should be 2020

File: src/cis/fam_cis_client.cpp [hunk 5]
```diff
@@ -0,0 +1,786 @@
...
+    ::grpc::Status status = server->stub->generate_profile(&ctx, req, &res);
+#endif
+}
+/**
+ * Creates a Region in FAM
+ * @param name-Name of the reion to be created
+ * @param nbytes - size of the region
+ * @param permissions - Permission with which the region needs to be created
+ * @param redundancyLevel - Redundancy level of FAM
+ * @return - pointer to Fam_Region_Descriptor
```
Comment (mashoodk) [inline src/cis/fam_cis_client.cpp:144]: Change the description of what is being returned.

File: src/cis/fam_cis_client.cpp [hunk 6]
```diff
@@ -0,0 +1,786 @@
...
+            return;
+        }
+    } else {
+        throw Fam_Allocator_Exception(FAM_ERR_GRPC,
+                                      (status.error_message()).c_str());
+    }
+}
+
+/**
+ * Chenages the permission of a dataitem
```
Comment (mashoodk) [inline src/cis/fam_cis_client.cpp:373]: spellcheck changes

File: src/cis/fam_cis_client.h [hunk 7]
```diff
@@ -0,0 +1,172 @@
+/*
+ * fam_cis_clienth
```
Comment (mashoodk) [inline src/cis/fam_cis_client.h:2]: Spellcheck fam_cis_client.h

File: src/fam-api/fam.cpp [hunk 8]
```diff
@@ -877,13 +841,13 @@ void fam::Impl_::clean_fam_options() {
...
     uint64_t key = descriptor->get_key();
-    Fam_Region_Item_Info itemInfo;
+    Fam_Region_Item_Info info;
 
     if (key == FAM_KEY_UNINITIALIZED) {
-        itemInfo = famAllocator->check_permission_get_info(descriptor);
-        descriptor->bind_key(itemInfo.key);
-        descriptor->set_size(itemInfo.size);
-        descriptor->set_base_address(itemInfo.base);
+        info = famAllocator->check_permission_get_info(descriptor);
```
Comment (mashoodk) [inline src/fam-api/fam.cpp:847]: May be set descriptor status as per new descriptor status.

File: src/allocator/fam_allocator_client.cpp [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+void Fam_Allocator_Client::resize_region(Fam_Region_Descriptor *descriptor,
+                                         uint64_t nbytes) {
+    Fam_Global_Descriptor globalDescriptor =
+        descriptor->get_global_descriptor();
+    uint64_t regionId = globalDescriptor.regionId;
+    uint64_t memoryServerId = descriptor->get_memserver_id();
+    try {
+        famCIS->resize_region(regionId, nbytes, memoryServerId, uid, gid);
```
Comment (mashoodk) [inline src/allocator/fam_allocator_client.cpp:107]: I think we should update descriptor statistics here.

File: src/cis/fam_cis_rpc.proto [hunk 10]
```diff
@@ -152,7 +153,8 @@ message Fam_Dataitem_Response {
     int32 errorcode = 6;
     string errormsg = 7;
     string name = 8;
-    uint64 perm = 9;
+	uint64 maxnamelen = 9;
```
Comment (mashoodk) [inline src/cis/fam_cis_rpc.proto:156]: alignment in proto files please

File: src/cis/fam_cis_rpc.proto [hunk 11]
```diff
@@ -1,6 +1,6 @@
 /*
- * fam_rpc.proto
- * Copyright (c) 2019 Hewlett Packard Enterprise Development, LP. All rights
+ * fam_cis_rpc.proto
+ * Copyright (c) 2020 Hewlett Packard Enterprise Development, LP. All rights
```
Comment (VrashiPonnappa) [inline src/cis/fam_cis_rpc.proto:3]:
```suggestion
 * Copyright (c) 2019-2020 Hewlett Packard Enterprise Development, LP. All rights
```

File: src/cis/fam_cis_server.h [hunk 12]
```diff
@@ -1,6 +1,6 @@
 /*
- * fam_rpc_service_impl.h
- * Copyright (c) 2019 Hewlett Packard Enterprise Development, LP. All rights
+ * fam_cis_server.h
+ * Copyright (c) 2020 Hewlett Packard Enterprise Development, LP. All rights
```
Comment (VrashiPonnappa) [inline src/cis/fam_cis_server.h:3]:
```suggestion
 * Copyright (c) 2019-2020 Hewlett Packard Enterprise Development, LP. All rights
```

File: src/common/fam_ops_shm.h [hunk 13]
```diff
@@ -1,5 +1,5 @@
 /*
- * fam_ops_nvmm.h
+ * fam_ops_shm.h
```
Comment (VrashiPonnappa) [inline src/common/fam_ops_shm.h:2]: update copyright

**Pull request merged**